File: landa/water_body_management/stocking_controller.py
import frappe
from frappe import _
from frappe.model.document import Document

from landa.utils import get_current_member_data

EARLIEST_YEAR = 2000
LATEST_YEAR = 2100


class StockingController(Document):
	def before_validate(self):
		self.set_weight_per_size()
		self.set_quantity_per_size()
		self.set_total_price()

	def validate(self):
		self.validate_year()
		self.validate_own_regional_organization()
		self.validate_own_water_body()

	def validate_year(self):
		if not EARLIEST_YEAR < self.year < LATEST_YEAR:
			frappe.throw(msg=_("Year must be between {0} and {1}.").format(EARLIEST_YEAR, LATEST_YEAR))

	def validate_own_regional_organization(self):
		member_data = get_current_member_data()
		if member_data and member_data.regional_organization != self.organization:
			frappe.throw(_("Please select your own regional Organization."))

	def validate_own_water_body(self):
		if (
			self.water_body
			and frappe.db.get_value("Water Body", self.water_body, "organization") != self.organization
		):
			frappe.throw(_("Please select a Water Body from your own regional Organization."))

	def set_weight_per_size(self):
		if not (self.weight and self.water_body_size):
			return

		self.weight_per_water_body_size = self.weight / self.water_body_size
		self.unit_of_weight_per_water_body_size = f"Kg / {self.water_body_size_unit}"

	def set_quantity_per_size(self):
		if not (self.quantity and self.water_body_size):
			return

		self.quantity_per_water_body_size = self.weight / self.water_body_size
		self.unit_of_quantity_per_water_body_size = f"Stk / {self.water_body_size_unit}"

	def set_total_price(self):
		if not (self.weight and self.price_per_kilogram):
			self.price_for_total_weight = 0
		else:
			self.price_for_total_weight = self.weight * self.price_per_kilogram

File: landa/water_body_management/stocking_controller.js
frappe.provide("landa.StockingController");

landa.StockingController = frappe.ui.form.Controller.extend({
    onload() {
		this.frm.set_query("organization", function (doc) {
			return {
				filters: {
					parent_organization: "LV",
					name: frappe.boot.landa.regional_organization,
				},
			};
		});
		this.frm.set_query("water_body", function (doc) {
			return {
				filters: {
					organization: doc.organization,
				},
			};
		});

		if (!this.frm.doc.organization && frappe.boot.landa.regional_organization) {
			this.frm.set_value("organization", frappe.boot.landa.regional_organization);
		}
	},
	water_body() {
		this.weight();
		this.quantity();
	},
    weight() {
		if (!this.frm.doc.weight || !this.frm.doc.water_body_size) {
			this.frm.set_value("weight_per_water_body_size", 0.0);
		} else {
			this.frm.set_value(
				"weight_per_water_body_size",
				flt(this.frm.doc.weight / this.frm.doc.water_body_size, 3)
			);
			this.frm.set_value(
				"unit_of_weight_per_water_body_size",
				`Kg / ${this.frm.doc.water_body_size_unit}`
			);
		}

		this.update_price();
	},
	quantity() {
		if (!this.frm.doc.quantity || !this.frm.doc.water_body_size) {
			this.frm.set_value("quantity_per_water_body_size", 0);
		} else {
			this.frm.set_value(
				"quantity_per_water_body_size",
				flt(this.frm.doc.quantity / this.frm.doc.water_body_size, 3)
			);
			this.frm.set_value(
				"unit_of_quantity_per_water_body_size",
				`Stk / ${this.frm.doc.water_body_size_unit}`
			);
		}
	},
	price_per_kilogram() {
		this.update_price();
	},
	update_price() {
		if (!this.frm.doc.weight || !this.frm.doc.price_per_kilogram) {
			this.frm.set_value("price_for_total_weight", 0);
		} else {
			this.frm.set_value(
				"price_for_total_weight",
				flt(this.frm.doc.weight * this.frm.doc.price_per_kilogram, 2)
			);
		}
	},
});